addons/mail/models/res_users.py

```diff
@@ -147,13 +143,33 @@ def systray_get_activities(self):
                     }
                 ],
             }
-            for record, activities in activities_by_record.items():
-                if record not in allowed_records:
-                    continue
-                for activity in activities:
-                    user_activities[model_name]["%s_count" % activity.state] += 1
-                    if activity.state in ("today", "overdue"):
-                        user_activities[model_name]["total_count"] += 1
+            query = """SELECT count(*),
+            CASE
+                WHEN %(today)s::date - act.date_deadline::date = 0 Then 'today_count'
+                WHEN %(today)s::date - act.date_deadline::date > 0 Then 'overdue_count'
+                WHEN %(today)s::date - act.date_deadline::date < 0 Then 'planned_count'
+            END AS states
+            FROM mail_activity AS act
+            WHERE user_id = %(user_id)s and res_id in %(allowed_records)s and res_model_id = %(res_model_id)s
+            GROUP BY states;
```

Comment:
Wouldn't it be possible to make this query for all activities only once, and to manually craft the WHERE clause for each model to inject it in?

But what part of this search actually requires a query to be more efficient than the ORM? Is it the grouping by state? Is it the count itself, like incrementing a variable in python is that slow?